liquid-compiler/src/parser.rs

```diff
@@ -194,12 +220,27 @@ fn parse_filter(filter: Pair, options: &Language) -> Result<FilterCall> {
             Error::with_msg("Unknown filter")
                 .context("requested filter", name.to_owned())
                 .context("available filters", available)
-        })?
-        .clone();
-    let f = FilterCall::new(name, f, args);
+        })?;
+    let f = f.parse(args)?;
+    
     Ok(f)
 }
 
+// TODO change HashMap and Vec into wrapped structs
+pub struct FilterArguments<'a> {
+    pub positional: Box<Iterator<Item=Expression>>,
+    pub keyword: HashMap<&'a str, Expression>,
+}
+impl<'a> FilterArguments<'a> {
+    pub fn check_args_exhausted(mut self) -> Result<()> {
```

Comment:
Do we need this?  Should this check just be in the macro or macro helpers?  This doesn't help us with keyword arguments.

Reply:
It helps as long as they are accessed with `remove`. I've been thinking about removing the `pub` on keyword and positional, and instead have two methods like `next_positional()` and `get_keyword(&str)` so it could be more clear and straightforward what methods to use to access the args.

It could be done in the macro, but in a way I intended to mimic `TagTokenIter::expect_nothing()`. If a user chooses not to use the macro (for whatever reason needs to manually write `FilterParameters::new()`), it would then still be clear that he couldn't forget about that function. I don't think this is gonna be a likely scenario, though, so I can just take it into the macro if you find that is the best option.